python/openai/tests/test_tool_calling.py

```diff
@@ -0,0 +1,571 @@
+            logprobs=False,
+            stream=True,
+        )
+
+        async for chunk in stream:
+            assert chunk.choices[0].index == 0
+
+            if chunk.choices[0].finish_reason:
+                finish_reason_count += 1
+                assert chunk.choices[0].finish_reason != "tool_calls"
+
+            # if a role is being streamed make sure it wasn't already set to
+            # something else
+            if chunk.choices[0].delta.role:
+                assert not role_name or role_name == "assistant"
+                role_name = "assistant"
+
+            # if a tool call is streamed make sure there's exactly one
+            # (based on the request parameters
+            streamed_tool_calls = chunk.choices[0].delta.tool_calls
+
+            if streamed_tool_calls and len(streamed_tool_calls) > 0:
+                assert len(streamed_tool_calls) == 1
+                tool_call = streamed_tool_calls[0]
+
+                # if a tool call ID is streamed, make sure one hasn't been already
+                if tool_call.id:
+                    assert not tool_call_id
+                    tool_call_id = tool_call.id
+
+                # if parts of the function start being streamed
+                if tool_call.function:
+                    # if the function name is defined, set it. it should be streamed
+                    # IN ENTIRETY, exactly one time.
+                    if tool_call.function.name:
+                        assert isinstance(tool_call.function.name, str)
+                        function_name = tool_call.function.name
+                    if tool_call.function.arguments:
+                        assert isinstance(tool_call.function.arguments, str)
+                        function_args_str += tool_call.function.arguments
+
+        assert finish_reason_count == 1
+        assert role_name == "assistant"
+
+        # validate the name and arguments
+        assert function_name == WEATHER_FORECAST_TOOL["function"]["name"]
+        assert function_name == tool_calls[0].function.name
+        assert isinstance(function_args_str, str)
+
+        # validate arguments
+        streamed_args = json.loads(function_args_str)
+        assert isinstance(streamed_args, Dict)
+        assert isinstance(streamed_args.get("city"), str)
+        assert isinstance(streamed_args.get("state"), str)
+        assert isinstance(streamed_args.get("unit"), str)
+        assert isinstance(streamed_args.get("num_days"), int)
+        assert streamed_args.get("city") == "Dallas"
+        assert streamed_args.get("state") == "TX"
+        assert streamed_args.get("unit") == "fahrenheit"
+
+        # make sure everything matches non-streaming except for ID
+        assert function_name == tool_calls[0].function.name
+        assert choice.message.role == role_name
+        assert choice.message.tool_calls[0].function.name == function_name
+
+    @pytest.mark.asyncio
+    async def test_tool_call_with_required_tool_choice(
```

Comment:
Do we have any negative tests like requesting or requiring a tool that doesn't exist?